### PR TITLE
Added support for `git stash list`

### PR DESCRIPTION
**Pull request by AlphabetsAlphabets**
Added basic support for `git stash list`. Current functionality in the `:Neogit` screen.
1. `Z` then `l` will give a list of stashes.
2. Pressing `ENTER` on any of the stashes will bring up a diff for that stash.

### CODE REVIEW

File: lua/neogit/buffers/stash_list_view/ui.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+local row = Ui.row
+
+-- local StashEntry = require("neogit.lib.git.stash").StashEntry
+
+local M = {}
+
+---Parses output of `git stash list` and splits elements into table
+M.Stash = Component.new(function(stash)
+  local children = {}
+  local table = table.concat({" stash@{", stash.idx, "}" }, "")
```
Comment (CKolkey) [inline lua/neogit/buffers/stash_list_view/ui.lua:16]:
```suggestion
  local lable = table.concat({" stash@{", stash.idx, "}" }, "")
```

File: lua/neogit/buffers/stash_list_view/ui.lua [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+local col = Ui.col
+local row = Ui.row
+
+-- local StashEntry = require("neogit.lib.git.stash").StashEntry
+
+local M = {}
+
+---Parses output of `git stash list` and splits elements into table
+M.Stash = Component.new(function(stash)
+  local children = {}
```
Comment (CKolkey) [inline lua/neogit/buffers/stash_list_view/ui.lua:15]:
```suggestion
```

File: lua/neogit/buffers/stash_list_view/ui.lua [hunk 3]
```diff
@@ -0,0 +1,39 @@
+local Ui = require("neogit.lib.ui")
+local Component = require("neogit.lib.ui.component")
+local util = require("neogit.lib.util")
+
+local text = Ui.text
+local col = Ui.col
+local row = Ui.row
+
+-- local StashEntry = require("neogit.lib.git.stash").StashEntry
```
Comment (CKolkey) [inline lua/neogit/buffers/stash_list_view/ui.lua:9]:
```suggestion
```

File: lua/neogit/buffers/stash_list_view/init.lua [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    mappings = {
+      n = {
+        ["q"] = function()
+          self:close()
+        end,
+        ["<esc>"] = function()
+          self:close()
+        end,
+        ["<enter>"] = function()
+          -- Still looking for how to view a stash
```
Comment (CKolkey) [inline lua/neogit/buffers/stash_list_view/init.lua:46]: You can use https://github.com/NeogitOrg/neogit/blob/f10084ae6ad75b6c7169dabde70d95e59a0b4c33/lua/neogit/lib/git/rev_parse.lua#L23 (`git rev-parse 'stash@{0}'`) to get the OID from a stash, which can be passed into the buffer.

File: lua/neogit/buffers/common.lua [hunk 5]
```diff
@@ -219,7 +219,7 @@ M.CommitEntry = Component.new(function(commit, args)
       }, graph, { text(" ") }, ref, { text(commit.subject) }),
       {
         virtual_text = {
-          { " ", "Constant" },
+          { " ",                                     "Constant" },
```
Comment (CKolkey) [inline lua/neogit/buffers/common.lua:222]:
```suggestion
          { " ", "Constant" },
```

File: lua/neogit/lib/git/stash.lua [hunk 6]
```diff
@@ -125,7 +130,13 @@ function M.register(meta)
     state.stashes.items = util.map(M.list(), function(line)
       local idx, message = line:match("stash@{(%d*)}: (.*)")
 
+      ---@class StashEntry
       return {
+        rel_date = cli.log
+          .max_count(1)
+          .format("%cr")
+          .args("stash@{" .. idx .. "}")
```
Comment (CKolkey) [inline lua/neogit/lib/git/stash.lua:138]:
```suggestion
          .args(("stash@{%s}"):format(idx))
```

Not a major thing, but in lua strings are interned, so concatenation operations will re-hash the new string, for every `..` operator. Many successive re-hashings will have a noticeable performance hit.  While it's probable that this list will never be long enough to encounter that, I just prefer to do string formatting or `table.concat()` to join strings :)

**Comment by CKolkey** [comment]
Nice work! Left some feedback

**Comment by AlphabetsAlphabets** [comment]
Hey, @CKolkey I fixed all the issues you brought up in the review. I also merged the latest from `master` and tested the code and all is well. Can you double check to make sure everything is alright?

**Comment by AlphabetsAlphabets** [comment]
@CKolkey hey there. Would you mind taking a look at the changes?

**Comment by CKolkey** [comment]
It's pretty good - there are still a handful of comments I left that are unresolved though. Take care of those and we'll merge this :)

**Comment by ej-shafran** [comment]
Hey - is there any movement on this? @AlphabetsAlphabets if you're no longer working on this feature, you mind if I fork off of you and fix the mentioned comments?

**Comment by AlphabetsAlphabets** [comment]
Hi, feel free to do so. I'm very busy right now.

**Pull request merged**